File: Solution209/Solution209.java
package Solution209;

/*
    209. Minimum Size Subarray Sum
    Given an array of positive integers nums and a positive integer target, return the minimal length of a subarray whose sum is greater than or equal to target. If there is no such subarray, return 0 instead.
    Example 1:

    Input: target = 7, nums = [2,3,1,2,4,3]
    Output: 2
    Explanation: The subarray [4,3] has the minimal length under the problem constraint.
*/
class Solution209 {
    public int minSubArrayLen(int target, int[] nums) {
        int left = 0;
        int currentSum = 0;
        int minLen = Integer.MAX_VALUE;

        for (int right = 0; right < nums.length; right++) {
            currentSum += nums[right];
            while (currentSum >= target) {
                minLen = Math.min(minLen, right - left + 1);
                currentSum -= nums[left];
                left++;        
            }
        }
        return (minLen == Integer.MAX_VALUE) ? 0 : minLen;
    }

    public static void main(String[] args) {
        Solution209 pro = new Solution209();
        int target = 7;
        int[] nums = {2, 3, 1, 2, 4, 3};
        int res = pro.minSubArrayLen(target, nums);
        System.out.println(res);
    }
}

File: Solution209/Solution209.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

class Solution209 {
public:
    int minSubArrayLen(int target,  vector<int>& nums) {
        int n = nums.size();
        int left = 0;
        int currentSum = 0;
        int minLen = INT_MAX;
        for (int right = 0; right<n; right++) {
            currentSum += nums[right];
            while (currentSum >= target) 
            {
                minLen = min(minLen, right-left+1);
                currentSum -= nums[left];
                left++;
            }
        }
        return (minLen == INT_MAX) ? 0 : minLen;
    }
};

int main() {
    Solution209 sol;
    int target = 7;
    vector<int> nums = {2, 3, 1, 2, 4, 3};

    cout << "Target: " << target << endl;
    cout << "Array: ";
    for (int x : nums) {
        cout << x << " ";
    }
    cout << endl;
    cout << "result: " << sol.minSubArrayLen(target, nums) << endl;
};